src/storage/kv/transaction.rs

```diff
@@ -132,31 +132,25 @@ pub struct Transaction {
     /// `mode` is the transaction mode. This can be either `ReadWrite`, `ReadOnly`, or `WriteOnly`.
     mode: Mode,
 
-    /// `snapshot` is the snapshot that the transaction is running in. This is a consistent view of the data at the time the transaction started.
+    /// `snapshot` is the snapshot that the transaction is running in. This is a consistent view of the data
+    ///  at the time the transaction started.
     pub(crate) snapshot: Option<RwLock<Snapshot>>,
 
-    /// `buf` is a reusable buffer for encoding transaction records. This is used to reduce memory allocations.
-    buf: BytesMut,
-
     /// `core` is the underlying core for the transaction. This is shared between transactions.
     pub(crate) core: Arc<Core>,
 
-    /// `write_order_map` is a mapping from sha256 of keys to their order in the write_set.
-    pub(crate) write_order_map: HashMap<Bytes, u32>,
-
-    /// `write_set` is a vector of  entries.
+    /// `write_set` is a map of keys to entries.
     /// These are the changes that the transaction intends to make to the data.
-    pub(crate) write_set: Vec<WriteSetEntry>,
+    /// The entries vec is used to keep different values for the same key for
+    /// savepoints and rollbacks.
+    pub(crate) write_set: HashMap<Bytes, Vec<WriteSetEntry>>,
```

Comment:
The intention of write_set being a vector was to preserve the order of writes

I am curious how this preserves the order in which writes are being written to the transaction. Let's say we have the following:

```
txn.set("k1", 1)
txn.set("k2", 2)
```

Could you help understand how would there be a guarantee k2 will be inserted after k1 here? The order will be important for deletes too if I think on the top of my head

Reply:
I'm actually not sure that the global order in needed. Let me also think about this over the weekend, but at the moment I do not think this is relevant, given that
- order per key is still preserved because of the vector, meaning that inserts and deletes for the same key are properly ordered
- global order between transactions is preserved with `commit_ts`, while in the same transactions all writes will have the same `commit_ts`.

Reply:
The sole reason to maintain order during initial implementation was for audit trails, to preserve the exact sequence of data insertion and the order of dependent operations, similar to a changelog. You are right, in the sense that not all types of applications might need this type of strict ordering, but I feel keeping the order preserved will be helpful in the long run. It is a safer bet, mainly because I can't think of all the scenarios right now :) but happy to discuss soon

Reply:
Let's extend the example above by adding another write for `k1`:
```
txn.set("k1", 1)
txn.set("k2", 2)
txn.set("k1", 3)
```
Now, with `write_order_map`, the write order will be [("k1", 3), ("k2", 2)], which is not the real time order that you'd expect, I think, because ("k1", 3) should come last. This happens because `write_order_map` remembers the _first_ time a key is written, and then it never changes the position even with additional writes later for the same key. I'm not sure that this property of ordering by the first write is particularly useful for auditing and thus worth preserving.

Reply:
You are right, this is a bug which needs fixing. Thanks for catching this. 

We will need to discuss and decide whether within a transaction we need to preserve this ordering or not, I see [badger](https://github.com/dgraph-io/badger/blob/499b1424621cc5730dab2315e3958643cb7131bc/txn.go#L261) also maintains a map, so they aren't doing it. In rocksdb, they seem to be just [appending](https://github.com/facebook/rocksdb/blob/0e04ef1a96d75acdcf2cf1802692375d62aea64c/db/write_batch.cc#L839) but I haven't looked too much into its details.

So we will need to decide if:

1) We get rid of ordering within a transaction
2) All three ( (k1, 1) (k2, 2) (k1, 3) ) get written
3) Any update to the previous written key updates the previous value ( (k1, 3) (k2, 2) )
4) Any update to the previous written key updates the last value ( (k2, 2) (k1, 3) )

Reply:
The write order within a transaction is now preserved using the write sequence numbering.